src/drivers/Unix/Fl_Unix_System_Driver.cxx

```diff
@@ -443,87 +444,135 @@ void Fl_Unix_System_Driver::newUUID(char *uuidBuffer)
-  if ((pref_type) != Fl_Preferences::USER)
-    return filename;
+/*
+ The user path changed between FLTK 1.3 and 1.4. It is now calculated
+ using XDG guidelines. It is `$XDG_CONFIG_HOME/{vendor}/{application}.prefs`
+ if `$XDG_CONFIG_HOME` is set, and `$HOME/.config/{vendor}/{application}.prefs`
+ if `$XDG_CONFIG_HOME` is not set or empty.
 
-  // If the legacy file exists, we are also done
-  if (::access(filename, F_OK)==0)
-    return filename;
+ For compatibility with 1.3 preferences, this function checks
 
-  // This is USER mode, and there is no legacy file. Create an XDG conforming path.
-  // Check $XDG_CONFIG_HOME, and if it isn't set, default to $HOME/.config
-  const char *xdg = getenv("XDG_CONFIG_HOME");
-  if (xdg==NULL) {
-    xdg = "~/.config";
+ 1. Does the XDG based top level folder '{vendor}' exist? If yes, use it.
+ 2. If not: does the old location $HOME/.fltk/{vendor} exist? If yes, use it.
+ 3. If neither: fall back to (1.) and use the XDG based folder (create it and 
+    the prefs file below it).
+ */
+char *Fl_Unix_System_Driver::preference_user_rootnode(
+    const char *vendor, 
+    const char *application, 
+    char *buffer) 
+{
+  // Find the path to the user's home directory.
+  Fl_String home_path = getenv("HOME");
+  if (home_path.empty()) {
+    struct passwd *pw = getpwuid(getuid());
+    if (pw)
+      home_path = pw->pw_dir;
   }
-  filename[0] = 0;
-  if (strncmp(xdg, "~/", 2)==0) {
-    strlcpy(filename, home, FL_PATH_MAX);
-    strlcat(filename, "/", FL_PATH_MAX);
-    strlcat(filename, xdg+2, FL_PATH_MAX);
-  } else if (strncmp(xdg, "$HOME/", 6)==0) {
-    strlcpy(filename, home, FL_PATH_MAX);
-    strlcat(filename, "/", FL_PATH_MAX);
-    strlcat(filename, xdg+6, FL_PATH_MAX);
-  } else if (strncmp(xdg, "${HOME}/", 8)==0) {
-    strlcpy(filename, home, FL_PATH_MAX);
-    strlcat(filename, "/", FL_PATH_MAX);
-    strlcat(filename, xdg+8, FL_PATH_MAX);
+
+  // 1: Generate the 1.4 path for this vendor and application.
+  Fl_String prefs_path_14 = getenv("XDG_CONFIG_HOME");
+  if (prefs_path_14.empty()) {
+    prefs_path_14 = home_path + "/.config";
   } else {
-    strlcpy(filename, xdg, FL_PATH_MAX);
+    if (prefs_path_14[prefs_path_14.size()-1]!='/')
+      prefs_path_14.append('/');
+    if (prefs_path_14.find("~/")==0) // starts with "~" 
+      prefs_path_14.replace(0, 1, home_path);
+    size_t h_env = prefs_path_14.find("${HOME}");
+    if (h_env!=prefs_path_14.npos)
+      prefs_path_14.replace(h_env, 7, home_path);
+    h_env = prefs_path_14.find("$HOME/");
+    if (h_env!=prefs_path_14.npos)
+      prefs_path_14.replace(h_env, 5, home_path);
+  }
+  if (prefs_path_14[prefs_path_14.size()-1]!='/')
+    prefs_path_14.append('/');
+  prefs_path_14.append(vendor);
+
+  // 2: If this base path does not exist, try the 1.3 path
+  if (::access(prefs_path_14.c_str(), F_OK) == -1) {
+    Fl_String prefs_path_13 = home_path + ".fltk/" + vendor;
+    if (::access(prefs_path_14.c_str(), F_OK) == 0) {
```

Comment:
s/prefs_path_14/prefs_path_13/ (?) (!)

Untested (yet)

Reply:
Yes, that one would have been an issue!